Store warnings in a set rather than a list

andreasabel opened the pull request:
Currently we store warnings in a list and have some hacks to prevent duplicate warnings.
Also, to join warning lists we use `List.union` which has a weird semantics.
It would be more natural to use a `Set` for warnings.

Currently we lack an `Ord` instance for `TCWarning` since `Doc` does not have one.
So here I cache the rendering of the warning doc as `tcWarningString` to facilitate an `Ord` instance.
The warnings should be ordered by position so the first component in the ordering is the `Range` of the warning.

~~WIP: For reasons very unclear to me, this first component is ignored, or the ranges there are mysteriously empty, so the warnings are ordered by the string rendering of their position.
I already tried to make sure that the ranges are serialized, but to no avail.
I need some help in debugging this.~~ Thanks @plt-amy !

New features:
- print range also for UnsolvedMeta and UnsolvedConstraints warnings (the latter often does not have a range, though)
- separate NonFatalErrors by newlines

andreasabel committed:
Add Ord TCWarning via cached String rendering

andreasabel committed:
Store warnings in a Set rather than a list

This ensures that warnings we register twice are not printed twice.
Also, the warnings are printed in order of their range.

andreasabel commented on test/Succeed/Generalize.warn:
```diff
@@ -1,3 +1,15 @@
+
+Generalize.agda:129,18-19: warning: -W[no]HiddenGeneralize
+Declaring a variable as hidden has no effect in a variable block.
+Generalization never introduces visible arguments.
```
An example where a warning at line 129 comes before one on 64 (textual rather than numerical ordering).

andreasabel replied:
@plt-amy offered a valid explanation: wrong ranged for Nificier warnings.

andreasabel commented on src/full/Agda/TypeChecking/Monad/Base.hs:
```diff
@@ -4625,9 +4629,11 @@ tcWarningOrigin = rangeFile . tcWarningRange
+  (==) = (==) `on` tcWarningRange &&& tcWarningString
+
+instance Ord TCWarning where
+  compare = compare `on` tcWarningRange &&& tcWarningString
```
This intends to order warnings by range first and then by the warning message.

gallais commented on src/full/Agda/Interaction/Imports.hs:
```diff
@@ -717,9 +717,9 @@ getStoredInterface x file msrc = do
+reportWarningsForModule :: MonadDebug m => TopLevelModuleName -> Set TCWarning -> m ()
 reportWarningsForModule x warns = do
-  unlessNull (filter ((Strict.Just (Just x) ==) . fmap rangeFileName . tcWarningOrigin) warns) \ ws ->
+  unlessNull (filter ((Strict.Just (Just x) ==) . fmap rangeFileName . tcWarningOrigin) $ Set.toList warns) \ ws ->
```
```suggestion
  unlessNull (Set.filter ((Strict.Just (Just x) ==) . fmap rangeFileName . tcWarningOrigin) warns) \ ws ->
```

andreasabel replied:
Thanks for the review @gallais!

So here I frankly do not know what is better, `List.filter` or `Set.filter`.  The result `ws` is then only used for printing, iterated through, so the list structure is the most natural there.  
Since `Set.filter` needs to produce a balanced tree, there might be some wasted effort of rebalancing.  In the worst case, `Set.filter f` is `Set.fromList . List.filter f . Set.toList`.

gallais replied:
Ah yes, my bad: I did not notice `unlessNull` actually had a continuation to run.

gallais commented on test/Fail/MdSpanningComment.err:
```diff
@@ -1,4 +1,4 @@
 error: [OverlappingTokensWarning]
-MdSpanningComment.lagda.md:2,4-6,19: Multi-line comment spans one or more literate text blocks.
-error: [OverlappingTokensWarning]
 MdSpanningComment.lagda.md:10,3-16,7: Multi-line comment spans one or more literate text blocks.
+error: [OverlappingTokensWarning]
+MdSpanningComment.lagda.md:2,4-6,19: Multi-line comment spans one or more literate text blocks.
```
This is strange: are errors now ordered in reverse location order?
Edit: Ah, my bad I see you mention it in the WIP part of the original post!

andreasabel commented on src/full/Agda/TypeChecking/Warnings.hs:
```diff
@@ -104,7 +99,7 @@ warning'_ loc w = do
     , prettyTCM c
     ]
-  return $ TCWarning loc r w p b
+  return $ TCWarning loc r w p (P.render p) b
```
Store `r'` here instead `r`.

andreasabel commented on src/full/Agda/TypeChecking/Errors.hs:
```diff
@@ -143,7 +143,7 @@ instance PrettyTCM TCErr where
     TypeError loc _ Closure{ clValue = NonFatalErrors ws } -> do
       reportSLn "error" 2 $ "Error raised at " ++ prettyShow loc
-      foldr1 ($$) $ fmap prettyTCM ws
+      vcat $ map prettyTCM $ Set.toAscList ws
```
Check whether we need `vsep` here (see #6919).

andreasabel replied:
Yep!

andreasabel commented on src/full/Agda/TypeChecking/Errors.hs:
```diff
@@ -1248,7 +1248,7 @@ instance PrettyTCM TypeError where
       pwords "Multiple polarity pragmas for" ++ map pretty xs
 
-    NonFatalErrors ws -> foldr1 ($$) $ fmap prettyTCM ws
+    NonFatalErrors ws -> vcat $ map prettyTCM $ Set.toAscList ws
```
Check whether we need `vsep` here (see #6919).

andreasabel replied:
Yep!

andreasabel commented on src/full/Agda/TypeChecking/Pretty/Warning.hs:
```diff
@@ -250,11 +250,11 @@ prettyWarning = \case
-    SafeFlagPragma xs -> fsep $ concat
-      [ [ fsep $ pwords "Cannot set OPTIONS" ++ [ pluralS (words =<< xs) "pragma" ] ]
+    SafeFlagPragma sset -> vcat $ concat
+      [ [ fwords $ singPlural (words =<< xs) id (++ "s") "The --safe mode does not allow OPTIONS pragma" ]
```
Check whether this rewording isn't a regression.

andreasabel commented on src/full/Agda/TypeChecking/Pretty/Warning.hs:
```diff
@@ -690,15 +690,15 @@ tcWarningsToError mws = case (unsolvedHoles, otherWarnings) of
+  let sfp = case foldMap pragmas ws of
+              (TCWarning loc r w doc str b : _, sfp) ->
+                 singleton $ TCWarning loc r (SafeFlagPragma sfp) doc str b
+              _ -> empty
```
Can we get rid entirely of this special logic for `SafeFlagPragma`?  One of these possibilities:
1. Make `SafeFlagPragma` take a single pragma name rather than a Set.
2. Store the `SafeFlagPragma` warning separately so that in the state these warnings are always properly aggregated.

Seems like 1. would be simplest.
Ad 2: We might want to aggregate other warnings as well (at least those that are deserialized), so we should have a uniform mechanism for that rather than this special case for `SafeFlagPragma`.

andreasabel commented on src/full/Agda/TypeChecking/Serialise/Instances/Errors.hs:
```diff
@@ -32,8 +33,8 @@ instance EmbPrj IsAmbiguous where
-  icod_ (TCWarning fp a b c d) = icodeN' TCWarning fp a b c d
-  value = valueN TCWarning
+  icod_ (TCWarning fp r a b c d) = icodeN' (\ fp -> TCWarning fp . underlyingRange) fp (SerialisedRange r) a b c d
+  value = valueN (\ fp -> TCWarning fp . underlyingRange)
```
Bump the interface version.

andreasabel commented on src/full/Agda/TypeChecking/Warnings.hs:
```diff
@@ -104,7 +99,7 @@ warning'_ loc w = do
     , prettyTCM c
     ]
-  return $ TCWarning loc r w p b
+  return $ TCWarning loc r' w p (P.render p) b
```
Consider only rendering only the actual warning `w` without the context (`c`, `r'`).  This might require ensuring that each warning actually comes with a range, otherwise different instances of the same warning will be lumped together.

andreasabel commented on test/Fail/Issue1375-2.err:
```diff
@@ -1,11 +1,11 @@
+Unsolved interaction metas at the following locations:
+  Issue1375-2.agda:24,13-17
+  Issue1375-2.agda:25,50-54
+  Issue1375-2.agda:26,21-25
```
The warning about unsolved metas should come with a range!

andreasabel commented on test/Fail/Issue4147.err:
```diff
@@ -3,8 +3,3 @@ Global confluence check failed: A unfolds to ⊥ which should further
-Global confluence check failed: A unfolds to ⊥ which should further
-unfold to ⊤ but it does not.
-Possible fix: add a rule to rewrite ⊥ to ⊤
-when checking the pragma REWRITE q
```
Good, removed a duplicate warning.

andreasabel commented on test/Fail/Issue4172-2.err:
```diff
@@ -8,16 +8,6 @@ Reason: It relies on injectivity of the data constructor false,
-        which is not yet supported
-
-when checking the definition of F
-
```
Removed duplicate.

andreasabel commented on test/Fail/Issue5029.err:
```diff
@@ -1,6 +1,3 @@
-Issue5029.agda:1,1-45: warning: -W[no]WarningProblem
-You may only turn off benign warnings. The warning TerminationIssue is a non-fatal error and thus cannot be ignored. See --help=warning.
-
```
Removed duplicate.
Missing newline between warning and error.

andreasabel replied:
> Missing newline between warning and error.

Cannot find the code position where to fix this.

andreasabel commented on test/Fail/Issue5681.err:
```diff
@@ -2,9 +2,9 @@ error: [UnsolvedConstraints]
   Issue5681.agda:3,14-18
+error: [UnsolvedMetaVariables]
+Unsolved metas at the following locations:
+  Issue5681.agda:3,8-11
```
Missing newlines.

andreasabel commented on test/Fail/MdSpanningComment.err:
```diff
@@ -1,4 +1,4 @@
-error: [OverlappingTokensWarning]
 MdSpanningComment.lagda.md:10,3-16,7: Multi-line comment spans one or more literate text blocks.
+error: [OverlappingTokensWarning]
+MdSpanningComment.lagda.md:2,4-6,19: Multi-line comment spans one or more literate text blocks.
```
Missing range for this error.

andreasabel replied:
Attempt to fix this in #7492.

andreasabel committed:
Refactor the checks isBuiltinModule and isPrimitiveModule

If `f` has inverse `g`, then check `g x ∈ s` is often faster then `x ∈ fmap f s`.

andreasabel committed:
Separate NonFatalErrors by an empty line

andreasabel committed:
Use non-empty sets to store UnsolvedConstraints in warning

Now they will be printed in order of their Range.

andreasabel committed:
Print Range for Unsolved* warnings
